Add 'llm' command snippet for MongoDB shell

juananpe opened the pull request:
This PR adds a new **llm** snippet command for the MongoDB shell (mongosh). The snippet provides helpful suggestions and recommendations for MongoDB-related tasks using the Groq LLM API (by default) or any model available locally in ollama. Key features include:

- Adds an 'llm' command to mongosh
- Generates query results using Groq API or any other locally available LLM (via Ollama)
- Provides brief, command-focused answers to MongoDB queries
- Easy installation using the 'snippet install' command
- Includes usage examples and troubleshooting tips

The snippet aims to enhance the user experience in mongosh by offering quick, relevant suggestions for common MongoDB operations and queries entered using natural language.

Examples:
(using Groq API)
```javascript
llm("very briefly, just the command, do not use markdown: in mongosh how to get the collections names of current db?");

db.getCollectionNames()
```

```javascript
llm("very briefly, just the command, do not use markdown: in mongosh replace all documents of a collection with property {'set':'llm102'} with the new value {'set':'llm101'} in current db?")

db.collection.updateMany({ set: 'llm102' }, { $set: { set: 'llm101' } })
```
(using Ollama+Phi3.5)
```javascript
llm("very briefly, just the command, do not use markdown: in mongosh how to get the collections names of current db?", {model: 'phi3.5'});

db.getCollectionNames()
```

See more:
https://github.com/juananpe/mongosh-snippets/blob/main/snippets/llm-command/README.md

Hope it helps.

juananpe committed:
feat: Add 'llm' command snippet

addaleax commented on package.json:
```diff
@@ -9,12 +9,12 @@
   },
   "bugs": {
-    "url": "https://github.com/mongodb-labs/mongosh-snippets/issues"
+    "url": "https://github.com/juananpe/mongosh-snippets/issues"
```
I think you might want to fix these 🙂

juananpe replied:
Fixed

addaleax commented on snippets/hint-command/package.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "@juananpe/snippet-hint-command",
+  "snippetName": "hint-command",
```
We've chatted about this a bit, and since it's currently tied to Groq as a service, we've come to the conclusion that it would make more sense to use `groq` as the snippet name/function name instead – what do you think about that?

juananpe replied:
I have added support for multiple local LLMs (any LLM available locally via Ollama)
If the user doesn't specify a model (optional second parameter), then Groq API will be used.
If the user hasn't provided a Groq API Key, the snippet will display a warning.
Thus, I've renamed the snippet as `llm` (instead of `hint` or `groq`)

addaleax commented on snippets/hint-command/package.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "@juananpe/snippet-hint-command",
```
This should match the other snippets w.r.t. the prefix, so that we can actually publish this snippet

juananpe replied:
Do you mean that I need to name it as @mongosh/snippet-llm-command?

addaleax commented on snippets/hint-command/index.js:
```diff
@@ -0,0 +1,25 @@
+        return "No message found in the response.";
+      }
+    } catch (error) {
+      console.error("Error fetching message:", error);
```
What does the error look like if `GROQ_API_KEY` is not set? Does it refer to the `GROQ_API_KEY` environment variable by name?

juananpe replied:
I have added this condition to warn the user if the groqApiKey is not set:

```
     // Groq (llama) version
        if (!groqApiKey) {
          console.warn('Groq API key is not set. Please set the GROQ_API_KEY environment variable.');
          return 'Unable to fetch message due to missing Groq API key.';
        }
```

addaleax reviewed (commented):
Hi @juananpe, thanks for the PR! Left a few comments, but I think generally we're up for including this in the official snippet list 👍

juananpe commented:
> Hi @juananpe, thanks for the PR! Left a few comments, but I think generally we're up for including this in the official snippet list 👍

Hi @addaleax, I have fixed the code based on your comments. I've added the option to select which LLM you want to use (Groq by default, or any other local LLM available in Ollama if the user prefers it). I've also squashed the commits into a single one. However, I have a doubt about this requirement: https://github.com/mongodb-labs/mongosh-snippets/pull/7#discussion_r1731826009 . Could you take a look at it? Thanks

addaleax commented on package.json:
```diff
@@ -9,10 +9,10 @@
   "repository": {
     "type": "git",
-    "url": "https://github.com/mongodb-labs/mongosh-snippets.git"
+    "url": "https://github.com/juananpe/mongosh-snippets.git"
```
```suggestion
    "url": "https://github.com/mongodb-labs/mongosh-snippets.git"
```

juananpe replied:
Done

addaleax commented on package.json:
```diff
@@ -9,10 +9,10 @@
     "show-index": "node scripts/show-index"
   },
-  "homepage": "https://github.com/mongodb-labs/mongosh-snippets",
+  "homepage": "https://github.com/juananpe/mongosh-snippets",
```
```suggestion
  "homepage": "https://github.com/mongodb-labs/mongosh-snippets",
```

juananpe replied:
Fixed

addaleax commented on snippets/llm-command/README.md:
````diff
@@ -0,0 +1,65 @@
+```
+export GROQ_API_KEY=gsk_XXXXXXX
+mongosh
+config.set('snippetIndexSourceURLs', 'https://github.com/juananpe/mongosh-snippets/raw/main/index.bson.br;' + config.get('snippetIndexSourceURLs') )
````
```suggestion
```

juananpe replied:
Line removed

addaleax commented on snippets/llm-command/package.json:
```diff
@@ -0,0 +1,16 @@
+  },
+  "dependencies": {
+    "groq-sdk": "^0.5.0",
+    "node-fetch-commonjs": "*"
```
```suggestion
    "node-fetch-commonjs": "^3.3.2"
```

juananpe replied:
Version applied

addaleax commented on index.bson.br:
Just a heads up, this file currently points at your own npm package, but it will be overwritten anyway when this PR is accepted

juananpe replied:
Right, no problem.

addaleax reviewed (approved):
Thank you!

juananpe committed:
fix: applied suggested changes

addaleax commented:
@juananpe Sorry for the delay – this has been published now! 🚀

juananpe commented:
> @juananpe Sorry for the delay – this has been published now! 🚀

Thanks. Are you planning to add any CONTRIBUTORS.md-like file that I can point to?

addaleax commented:
@juananpe Not currently, but PRs are welcome for everything. Right now, the README of this repo serves as our main guide for how to contribute to this repository
